### PR TITLE
Adapted needlist for base_url rel path

### LINKED ISSUE #407: Support relative base_path in needs_external_needs
@danwos  I wrote this request a while ago in [this comment](https://github.com/useblocks/sphinxcontrib-needs/issues/339#issuecomment-888843531) in issue #339 but it should have it's own ticket.

Sphinx-needs doesn't support relative paths to an "external needs" project indeed. I just tested this.

This would come in handy for relocatable multi-project build documentation, i.e. if you put the documentation in such a web folder structure:

```
<webroot>
|- master
|  |- project_a
|  |- project_b  # reference project A
|  `  ...
|- release
|  |- 1.0
|  |  |- project_a
|  |  |- project_b  # reference project A
|  |  `  ...
|  |- 1.1
|  ` ...
```

Intersphinx supports this:

```
intersphinx_mapping = {'project_a': ('../../project_a/html',
                                  (os.path.join(project_root, 'build', 'project_a', 'html', 'objects.inv'), None))}
```

Result:

`:doc:` links will be converted to links in the following pattern:

`<relative path to my project root>/<relative path to other project root>/<relative document path>.html`

e.g. if you give have a role :doc:`project_a:tools/index` in a file `project_b/html/doc/design.rst`, the relative link will be:

`../../../project_a/html/tools/index.hml`

It now does not matter whether the documentation is in a relative path master/ or release/1.0/ or release/1.1/

I don't have an executable example but my config looks like this:

```
# config file of project_b
external_needs_json_file = os.path.join(project_root, 'build', 'project_a', 'needs', 'needs.json')
if os.path.isfile(external_needs_json_file):
    needs_external_needs = [
        {
            "base_url": "../../project_a/html",
            "json_path": external_needs_json_file,
            "css_class": "external_link",
        },
    ]

# output of project_b will be in <project_root>/build/project_b/html
```

While intersphinx actually keeps ".." in relative paths, needs_external_needs seems to simply remove it and replace relative links by absolute links.

### PR DESCRIPTION
**Pull request by haiyangToAI**
Fix #407

### CODE REVIEW

**Comment by arwedus** [comment]
first test on my projects looks promising 👍🏼

File: sphinxcontrib/needs/diagrams_common.py [hunk 1]
```diff
@@ -175,11 +177,23 @@ def calculate_link(app, need_info):
...
         if need_info["is_external"]:
             link = need_info["external_url"]
+
+            # check if given base_url is url or relative path
+            parsed_url = urlparse(need_info["external_url"])
+            if not parsed_url.scheme and not os.path.isabs(need_info["external_url"]):
+                # get path sep considering plattform dependency, '\' for Windows, '/' fro Unix
+                curr_path_sep = os.path.sep
+                # check / or \ to determine the relative path to conf.py directory
+                if curr_path_sep in fromdocname:
```
Comment (danwos) [inline sphinxcontrib/needs/diagrams_common.py:193]: if -statement could be combined with above if statement

File: sphinxcontrib/needs/directives/needlist.py [hunk 2]
```diff
@@ -103,7 +105,18 @@ def process_needlist(app, doctree, fromdocname):
...
                 ref = nodes.reference("", "")
+
                 ref["refuri"] = need_info["external_url"]
+                # check if given base_url is url or relative path
+                parsed_url = urlparse(need_info["external_url"])
+                if not parsed_url.scheme and not os.path.isabs(need_info["external_url"]):
+                    # get path sep considering plattform dependency, '\' for Windows, '/' fro Unix
+                    curr_path_sep = os.path.sep
+                    # check / or \ to determine the relative path to conf.py directory
+                    if curr_path_sep in fromdocname:
```
Comment (danwos) [inline sphinxcontrib/needs/directives/needlist.py:116]: same, here, one if-statement?

File: sphinxcontrib/needs/roles/need_incoming.py [hunk 3]
```diff
@@ -53,6 +56,19 @@ def process_need_incoming(app, doctree, fromdocname):
...
                         new_node_ref = nodes.reference(target_need["id"], target_need["id"])
                         new_node_ref["refuri"] = target_need["external_url"]
+
+                        # check if given base_url is url or relative path
+                        parsed_url = urlparse(target_need["external_url"])
+                        if not parsed_url.scheme and not os.path.isabs(target_need["external_url"]):
+                            # get path sep considering plattform dependency, '\' for Windows, '/' fro Unix
+                            curr_path_sep = os.path.sep
+                            # check / or \ to determine the relative path to conf.py directory
+                            if curr_path_sep in fromdocname:
```
Comment (danwos) [inline sphinxcontrib/needs/roles/need_incoming.py:66]: again :)

File: sphinxcontrib/needs/roles/need_outgoing.py [hunk 4]
```diff
@@ -74,6 +77,19 @@ def process_need_outgoing(app, doctree, fromdocname):
                     else:
                         new_node_ref = nodes.reference(target_need["id"], target_need["id"])
                         new_node_ref["refuri"] = target_need["external_url"]
+
```
Comment (danwos) [inline sphinxcontrib/needs/roles/need_outgoing.py:80]: okay, we have the same code on different location.
That's hard to maintain, if we detect an error in it.
Then we need to fix it at >4 locations.

Any chance to create a common function for this? maybe in ``util.py``

**Review by danwos — CHANGES_REQUESTED** [review]

**Review by danwos — APPROVED** [review]
Looks all good.
Great work.

**Pull request merged**